<a href="https://colab.research.google.com/github/onnex2024/ucice/blob/main/M3_Assignment_HD_Colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Load the datasets "/content/DIA_trainingset_RDKit_descriptors.csv" and "/content/DIA_testset_RDKit_descriptors.csv", train a decision tree model on the training data, and evaluate it on the test data.

## Load data

### Subtask:
Load the training and testing datasets into pandas DataFrames.


**Reasoning**:
The first step is to import the pandas library and load the training and testing datasets into pandas DataFrames as instructed.



In [ ]:
import pandas as pd

train_df = pd.read_csv("/content/DIA_trainingset_RDKit_descriptors.csv")
test_df = pd.read_csv("/content/DIA_testset_RDKit_descriptors.csv")

print("Training data head:")
display(train_df.head())

print("\nTesting data head:")
display(test_df.head())

Training data head:


Label                                             SMILES  BalabanJ  \
0      0  COC(=O)N(C)c1c(N)nc(nc1N)c2nn(Cc3ccccc3F)c4ncc...     1.821   
1      0                    C[C@H](N(O)C(=O)N)c1cc2ccccc2s1     2.363   
2      0                              C[N+](C)(C)CC(=O)[O-]     3.551   
3      1  CC(C)n1c(\C=C\[C@H](O)C[C@H](O)CC(=O)O)c(c2ccc...     2.076   
4      1           C\C(=C(\C#N)/C(=O)Nc1ccc(cc1)C(F)(F)F)\O     2.888   

    BertzCT    Chi0   Chi0n   Chi0v    Chi1  Chi1n  Chi1v  ...  fr_sulfide  \
0  1266.407  22.121  16.781  16.781  14.901  9.203  9.203  ...           0   
1   490.434  11.707   8.752   9.569   7.592  4.854  5.670  ...           0   
2    93.092   6.784   5.471   5.471   3.417  2.420  2.420  ...           0   
3  1053.003  21.836  16.995  16.995  14.274  9.926  9.926  ...           0   
4   549.823  14.629   9.746   9.746   8.752  5.040  5.040  ...           0   

   fr_sulfonamd  fr_sulfone  fr_term_acetylene  fr_tetrazole  fr_thiazole  \
0             0           0                  0             0            0   
1             0           0                  0             0            0   
2             0           0                  0             0            0   
3             0           0                  0             0            0   
4             0           0                  0             0            0   

   fr_thiocyan  fr_thiophene  fr_unbrch_alkane  fr_urea  
0            0             0                 0        0  
1            0             1                 0        1  
2            0             0                 0        0  
3            0             0                 0        0  
4            0             0                 0        0  

[5 rows x 198 columns]


Testing data head:


Label                                             SMILES  BalabanJ  \
0      0  C[C@H](\C=C\[C@H](O)C1CC1)[C@@H]2CC[C@@H]3\C(=...     1.484   
1      1     OCCN1CCN(CCCN2c3ccccc3Sc4ccc(cc24)C(F)(F)F)CC1     1.472   
2      0  C[C@@H]1O[C@H](C[C@H](O)[C@@H]1O)O[C@@H]2[C@H]...     0.837   
3      1                    NC(=O)Cc1cccc(C(=O)c2ccccc2)c1N     2.406   
4      0  COc1cc2c(CCN[C@]23CS[C@@H]4[C@@H]5[C@@H]6N(C)[...     1.320   

    BertzCT    Chi0   Chi0n   Chi0v    Chi1   Chi1n   Chi1v  ...  fr_sulfide  \
0   743.207  21.466  18.764  18.764  14.292  12.106  12.106  ...           0   
1   868.947  21.140  16.736  17.553  14.453  10.268  11.084  ...           0   
2  1409.004  39.189  32.904  32.904  26.011  20.941  20.941  ...           0   
3   621.298  13.828  10.297  10.297   9.092   5.847   5.847  ...           0   
4  2127.996  37.955  30.849  31.666  25.910  18.066  19.115  ...           1   

   fr_sulfonamd  fr_sulfone  fr_term_acetylene  fr_tetrazole  fr_thiazole  \
0             0           0                  0             0            0   
1             0           0                  0             0            0   
2             0           0                  0             0            0   
3             0           0                  0             0            0   
4             0           0                  0             0            0   

   fr_thiocyan  fr_thiophene  fr_unbrch_alkane  fr_urea  
0            0             0                 0        0  
1            0             0                 0        0  
2            0             0                 0        0  
3            0             0                 0        0  
4            0             0                 0        0  

[5 rows x 198 columns]

## Prepare data

### Subtask:
Separate features and target variables for both training and testing datasets.


**Reasoning**:
Separate the features and target variables for both the training and testing datasets as instructed.



In [ ]:
X_train = train_df.drop(['Label', 'SMILES'], axis=1)
y_train = train_df['Label']

X_test = test_df.drop(['Label', 'SMILES'], axis=1)
y_test = test_df['Label']

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)
print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_train shape: (477, 196)
y_train shape: (477,)
X_test shape: (120, 196)
y_test shape: (120,)


## Train model

### Subtask:
Train a decision tree classifier model using the training data.


**Reasoning**:
Train a decision tree classifier model using the training data.



In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Instantiate a DecisionTreeClassifier object
dt_model = DecisionTreeClassifier(random_state=42)

# Train the decision tree model on the training data
dt_model.fit(X_train, y_train)

print("Decision Tree model trained successfully.")

Decision Tree model trained successfully.


## Evaluate model

### Subtask:
Evaluate the trained model using the testing data to assess its performance.


**Reasoning**:
Import the necessary evaluation metrics and make predictions on the test data.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred = dt_model.predict(X_test)

**Reasoning**:
Calculate and print the evaluation metrics.



In [ ]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.7250
Precision: 0.4444
Recall: 0.4000
F1-score: 0.4211


## Summary:

### Data Analysis Key Findings

*   The training dataset contained 477 samples and 198 columns (including 'Label' and 'SMILES'), while the testing dataset contained 120 samples and the same number of columns.
*   After separating features and the target variable, both the training and testing feature sets (`X_train` and `X_test`) had 196 columns, and the target variables (`y_train` and `y_test`) were single-column series.
*   A Decision Tree Classifier model was successfully trained on the training data.
*   The trained Decision Tree model achieved an accuracy of 0.7250 on the test set.
*   The precision of the model on the test set was 0.4444, the recall was 0.4000, and the F1-score was 0.4211.

### Insights or Next Steps

*   The current model performance, particularly the precision and recall, suggests potential areas for improvement. Exploring different machine learning algorithms or hyperparameter tuning for the Decision Tree could enhance the model's predictive capabilities.
*   Investigating feature importance from the trained Decision Tree could provide insights into which RDKit descriptors are most influential in predicting the 'Label', potentially informing feature selection or engineering for future iterations.
